### PR TITLE
Version 1.0.3

### PR DESCRIPTION
**Pull request by FriendsOfGalaxyBot**
Sync with the original repository

### CODE REVIEW

File: src/version.py
Comment (FriendsOfGalaxy) [inline src/version.py]: this is rather 0.9.0 or 1.0.0, isn't it?:)

File: src/consts.py
Comment (FriendsOfGalaxy) [inline src/consts.py:16]: Do you plan to support this version in the future?

File: src/local.py
Comment (FriendsOfGalaxy) [inline src/local.py:32]: Shouldn't be `os.path.dirname(exe)`?

File: src/local.py
Comment (FriendsOfGalaxy) [inline src/local.py:38]: It is hard to follow this logic. You may consider refactor in the future to make it more maintainable. I would limit purposes and simplify the method interface, or at least document it /cover with tests. 

But for now, better do not change anything if it works :) (I see that it is used in many places around the code)

File: src/local.py
Comment (FriendsOfGalaxy) [inline src/local.py]: probably bug? I see too many negations...

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: Better put the logic in `get_local_size` and skip  `prepare_local_size_context` at all.

```python

    async def get_local_size(self, game_id: str, context):
        size = await utils.get_size_at_path(
            self.local_client.find_launcher_path(game_id, folder=True)
        )
        if game_id == GameID.Minecraft and self._multimc_enabled():
            if size is None:
                size = 0
            size += await utils.get_size_at_path(self.multimc.folder)
```

File: src/page/page3.html
Comment (FriendsOfGalaxy) [inline src/page/page3.html]: It was already put on page2 - now it is too late. Add info how to run MultiMC instead (double click on Play)

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py:51]: By the way: have you consider fetch ownership from mojang account? https://www.minecraft.net/en-us/profile/ -> https://api.mojang.com/user/profiles/agent/minecraft

It is not a blocker for this release. Not having to worry about authentication is a good thing... I'm just putting an idea for more comprehensive solution and less config in the future.

File: src/utils.py
Comment (FriendsOfGalaxy) [inline src/utils.py]: blocking I/O may lead to plugin crash (Galaxy may kill plugin that does not respond to `ping` in 10 seconds). It will be OK until downloaded file is very small or `download` is run in separate thread:

```python
installer_path = await asyncio.get_running_loop().run_in_executor(None, utils.download, url)
```

Another way is to make it async for example by using `aiofiles` (but why using 3rd party lib?)

File: requirements.txt
Comment (FriendsOfGalaxy) [inline requirements.txt]: +1 for already resolved requirements for this plugin

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: API requires `GameTime` to be returned. It is ok to return `GameTime('mc', None, None)`

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: some people say closing files is important. Use `with open` if you agree

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: yet? :) plugin is going to shutdown

File: src/page/page2.html
Comment (FriendsOfGalaxy) [inline src/page/page2.html]: why is it important to write? I'm not familiar with MultiMC but doesn't it allow to install the game?

And - if Vanila Mc is not installed, why not to run MultiMC on clicking play? Ignore me if this is wrong idea that only complicates the code.

**Review by FriendsOfGalaxy — CHANGES_REQUESTED** [review]
@TouwaStar I'm linking @urwrstkn8mare as well because he is an author.

I like those changes especially nice game chooser 👍 
I've added some comments that requires attention, please take a look.

I have also a general note on UX:
Setting up MultiMC may be not achievable for newbies. The instruction on MultiMC is not intuitive - 1) there is zip (not installer) for Windows 2) it requires manual installation of Java (maybe because Minecraft required Jave in the past).

Maybe it would be good to add something like
"(i) MultiMC requires installing Java  (and link to https://www.java.com/download/)" 
on page2 or link better instruction?

File: src/page/page2.html
Comment (FriendsOfGalaxy) [reply src/page/page2.html]: Or maybe add possibility to download Vanila Minecraft just from page3 (only if not installed yet)? Just ideas - it does not block merging this - I'm just putting ideas, you know the best what should be done.

File: src/consts.py
Comment (urwrstkn8mare) [reply src/consts.py:16]: Yeah, that was what I was planning but since the education edition works very differently it would require new logic and I hadn't had time to get to that yet.

File: src/version.py
Comment (urwrstkn8mare) [reply src/version.py]: True!

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py:32]: Yeah sorry, the code for that is a little messy over the multiple classes. Basically, that line is for Mac when the subclass returns the path to `Minecraft.app`. The thing is that's both the executable and the folder! That's why it's necessary but I could probably clean it up a little.

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py:38]: Yeah I've touched on the badness of this code in this reply to one of your other comments ([here](https://github.com/FriendsOfGalaxy/galaxy-integration-minecraft/pull/8#discussion_r482821587)). Basically you are completely correct. It does work so I'm not touching it but definitely it would be better if a refactor it in the future.

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py]: Sorry, I'm not sure what you're talking about here. But I'll try to answer as best as I can. 

Looking at it I realise I could've written together but I'm not sure where the bug is. The `rp is not None` line checks if `self.running_games[game_id]` is None and thus has been assigned a process at some point (and yes, I could've done this in one if statement wrapping the two next statements). 

`rp.poll()` is just to see if the process returned when we launched the game is still running. I used this instead of psutils so we don't have to repeatedly iterate over a list of processes and try to find the game.

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: good point, dunno how I missed that. sorry

File: src/page/page3.html
Comment (urwrstkn8mare) [reply src/page/page3.html]: yes, that is a great idea!

File: src/utils.py
Comment (urwrstkn8mare) [reply src/utils.py]: Yes, that is a very true problem that I actually experienced when my internet was slow. I was hoping it wouldn't be a problem but that is a bug that definitely needs fixing :).

File: requirements.txt
Comment (urwrstkn8mare) [reply requirements.txt]: Lol, I can't accept credit for that. That was @TouwaStar!

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: oh ok sorry about that.

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: yes definitely that was a mistake on my part

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: I did that because it causes an exception otherwise. But I realise now I could just stop tracking the game manually as if the user has finished and then try again. Thanks for noticing that.

File: src/page/page2.html
Comment (urwrstkn8mare) [reply src/page/page2.html]: > why is it important to write? I'm not familiar with MultiMC but doesn't it allow to install the game?

MultiMC and Vanilla Minecraft both allow you to install the game, but they are still separate launchers. There's honestly not much difference if you have one installed (unless you want mods) but I just put that in just so they know. 

> why not to run MultiMC on clicking play?

I'm pretty sure that's already how it is. I use the plugin on my PC and I don't have the vanilla MC installed. I specifically wrote some logic to ensure it doesn't wait for a double click as well to save some time (even if a little :)). Sorry if I'm missing something please correct me if I'm wrong.

> maybe add possibility to download Vanila Minecraft just from page3

Great idea!

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py:51]: I'm not sure how to use that but I'll definitely have a look at that.

**Comment by urwrstkn8mare** [comment]
I've replied to I think all the comments in the code. Additionally in regards to the main comment about UX. That is a very good point and I have some ideas that I will probably use to fix this. 

Unfortunately similar to my [Riot plugin](https://github.com/urwrstkn8mare/galaxy-riot-integration) where I do have planned features/fixes, I am very very very short on time with other work that have to take priority :(. It is likely I will have a window of opportunity to work on the riot plugin and this plugin during the holidays (in approx. 3 weeks). 

Also, how would I go about making the changes you've requested? This plugin belongs to @TouwaStar. So do I work on it and submit a pull request to his repo? Or do I make the changes in this pull request? (actually, I'm not sure I _can_ make changes in this pull request). Sorry, just some confusion on how to go about this.

Overall thanks for the in-depth feedback!

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py:51]: Just found this: https://github.com/summer/mojang
I could use this to auto select Minecraft and show users name in the integrations window in settings. There doesn't appear to be a way to see if Minecraft dungeons is installed

**Comment by TouwaStar** [comment]
@urwrstkn8mare  i think the flow would be to create a pull request to my repo, i would accept it and do the review here with @FriendsOfGalaxy  . i think thats the smoothest

File: src/local.py
Comment (TouwaStar) [reply src/local.py]: i think this is a bit confusing ``and not rp.poll() is not None:`` shouldnt it be just ``and rp.poll() is not None:`` instead ?

File: src/version.py
Comment (TouwaStar) [reply src/version.py]: I've upped the version just one minor up but i completely agree. @urwrstkn8mare  feel free to upp it to 1.0.0 when adhering to comments.  (has to be upped both in version.py and manifest.json)

File: src/local.py
Comment (FriendsOfGalaxy) [reply src/local.py]: TouwaStar is right, `not poll()` is bool so it is always `not None`

- process is still running:
`poll() == None` -> `not poll() == True` -> `not poll() is not None == True`
- process stopped with 0 status code:
`poll() == 0` -> `not poll() == True` -> `not poll() is not None == True`
- process stopped with status code other than 0:
`poll() == 1` -> `not poll() == False` -> `not poll() is not None == True`

File: src/plugin.py
Comment (FriendsOfGalaxy) [reply src/plugin.py]: no problem I'm just helping :)

File: src/plugin.py
Comment (FriendsOfGalaxy) [reply src/plugin.py]: yes, that would be good workaround to "stop" session even if game is still running when plugin/Galaxy is closed.

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py]: Oh, I see the issue now 🤦. I think I can refactor it to something like this:
```python
def is_game_still_running(self, game_id) -> bool:
	return (self.running_games[game_id] is not None 
	        and self.running_games[game_id] is None)
```
There's no point of reassigning None as another process will just replace whatever the value is of the property `self.running_games[game_id]` when a game starts. I can simplify the code down to this basically.

Comment (FriendsOfGalaxy) [reply src/local.py]: yes, almost - you forget inspect `poll()`

```python
def is_game_still_running(self, game_id) -> bool:
    return self.running_games[game_id] and self.running_games[game_id].poll() is None
```

File: src/local.py
Comment (FriendsOfGalaxy) [reply src/local.py:32]: Ou, I see. In this case the best would be to inject OS implementation to LocalClient (composition) and expose only higher level methods in LocalClient (like `run` or whatever you need there). Or, to avoid refactor just put some "if" statements. It is better to have more code but explicite.

```python
if WINDOWS:
    do_this()
else:
    do_that()
```

Comment (FriendsOfGalaxy) [reply src/local.py:32]: every folder is executable BTW. "Running" .app folders on Mac is just a syntastic sugar for system layer to run proper executable in `x.app/Contents/MacOS`

File: src/local.py
Comment (FriendsOfGalaxy) [reply src/local.py:38]: Just a tip for the future: follow single responsibility rule. Keep logic in relevant layer. Do simple  interfaces: single method that do single simple thing like: `find_launcher_dir()` (without arguments that modify the inner logic). Such method should return always path to launcher and that's it. Then you can create `find_executable(directory, [game_id](if really needed))` to get exact executable and `run()` method separate for Mac and Windows that utilize other methods under the hood.

**Comment by FriendsOfGalaxy** [comment]
@urwrstkn8mare I'm really glad that you've done all those nice feature and I think people will like them.

I regret that I didn't respond quicker for this PR and your Riot plugin when you were on holidays :)

Maybe the way to go is to better document things to do and try to invite other community dev to your work?

**Comment by urwrstkn8mare** [comment]
@FriendsOfGalaxy Thanks! You're definitely right. I'm obviously open to anyone else contributing/helping out. But as you said, I think I have to better document things to do and in general so people can help.

So I have two questions: 
1. what do you mean by better document things to do? Like better document _the todo_? Or document something else like the integration? Because I know there is some stuff in there that is hard to read but I can explain them to anyone if they ask. 
2. By inviting another community dev to you work did you mean just being open to help like pull requests? Or actually picking someone to jointly maintain the integration with me? If so who and how do I pick?

Sorry, I think I know what you mean but I just wanted a little clarification just to make sure :)

PS. I should be able to start a little earlier than the holidays because the workload will start to lighten as I get nearer to the holidays. Basically, I'll try to get to this as fast as I can.

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py:32]: > In this case the best would be to inject OS implementation to LocalClient (composition) and expose only higher level methods in LocalClient (like `run` or whatever you need there).

Yes, I definitely need to refactor local client to be better, in fact I kind of started halfway with the `launch` functions but then stopped because it was working. I'll work on it anyway :)

> "Running" .app folders on Mac is just a syntastic sugar for system layer to run proper executable in `x.app/Contents/MacOS`.

Absolutely did not know that, thanks a lot for that info.

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py:38]: Thanks that's very true, I will be doing some heavy refactoring in local.py

File: src/local.py
Comment (urwrstkn8mare) [reply src/local.py]: 🤦

File: src/local.py
Comment (FriendsOfGalaxy) [reply src/local.py:38]: Please do not refactor until we merge current working state :)

**Comment by FriendsOfGalaxy** [comment]
> 1. what do you mean by better document things to do? Like better document _the todo_?

yes, but in the form of issues with short description what should be done

> Or document something else like the integration?

Document in readme how to use you own debug scripts/tools you've shipped here.
 
> 2. By inviting another community dev to you work did you mean just being open to help like pull requests?

yes

> Or actually picking someone to jointly maintain the integration with me? If so who and how do I pick?

I think that the question is mostly to @TouwaStar as maintainer.

Maybe he can add you as collaborator? https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/inviting-collaborators-to-a-personal-repository

Then you may define some internal rules like: pushes to master only via PR that should be approved by one more person.

**Comment by urwrstkn8mare** [comment]
Thanks @FriendsOfGalaxy, I will make sure to make my todos issues and documenting my scripts better (part of it was adapted from a part of your [integration updater](https://github.com/FriendsOfGalaxy/galaxy-integrations-updater)). These questions were for both this integration and my [Riot integration](https://github.com/urwrstkn8mare/galaxy-riot-integration) except for the last question but I guess it applies as well.

> Maybe he can add you as collaborator? docs.github.com/en/github/setting-up-and-managing-your-github-user-account/inviting-collaborators-to-a-personal-repository

I'm happy to be a collaborator but I'm ok with just opening pull requests as a contributor if that is what you prefer @TouwaStar.

> yes, but in the form of issues with short description what should be done

I can definitely do this for my [Riot integration](https://github.com/urwrstkn8mare/galaxy-riot-integration) but are you saying I also open an issue for anything I plan to do in @TouwaStar's repo?

> Document in readme how to use you own debug scripts/tools you've shipped here.

No problem, I'll do that in both integrations. (part of it was actually adapted from a part of your [integration updater](https://github.com/FriendsOfGalaxy/galaxy-integrations-updater))


Sorry, this is my first somewhat significant contribution to another open-source project and I obviously don't know a lot of stuff :)

**Comment by FriendsOfGalaxy** [comment]
A few more things to encourage other people to contribute:
- follow github recommendations: https://docs.github.com/en/github/building-a-strong-community/setting-up-your-project-for-healthy-contributions
- ask devs on GOG Cafe discord
- move project to dedicated organization instead of private project (like https://github.com/vim/vim). But I have seen only big projects to be maintained like that, so no idea if it can really help

**Comment by urwrstkn8mare** [comment]
@FriendsOfGalaxy Sorry just a heads up I made some pretty big changes to my [comment](https://github.com/FriendsOfGalaxy/galaxy-integration-minecraft/pull/8#issuecomment-688218509). Like a few seconds before you just replied. I also had one last question.

**Comment by FriendsOfGalaxy** [comment]
> I can definitely do this for my Riot integration but are you saying I also open an issue for anything I plan to do in @TouwaStar's repo?

hmm, that is up to TouwaStar or rather what you both agree that should be done.

> (part of it was actually adapted from a part of your integration updater)

ok. FYI there is also https://github.com/FriendsOfGalaxy/buildtools released on pypi that I'm going to use instead of spreaded code in multiple plugins.

> Sorry, this is my first somewhat significant contribution to another open-source project and I obviously don't know a lot of stuff :)

No worries, there are no strict rules but rather common practices or it depends on rules in particular project. This fork-based system of integration-updater is also very custom thing.

**Comment by urwrstkn8mare** [comment]
Thanks, the `buildtools` on PyPI will be very useful. Btw there is not much documentation but I think I can understand how to use it from looking at the code so it is not that much of a problem.

**Comment by urwrstkn8mare** [comment]
Ok, @FriendsOfGalaxy @TouwaStar, I've managed to make some time and I think I've made all the necessary changes. I've submitted the [PR](https://github.com/TouwaStar/Galaxy_Plugin_Minecraft/pull/10#issue-483298227) to @TouwaStar's [repo](https://github.com/TouwaStar/Galaxy_Plugin_Minecraft).

**Comment by urwrstkn8mare** [comment]
@FriendsOfGalaxy Just a probably unrelated question: Shouldn't @FriendsOfGalaxyBot only merge the `src` directory into this repo. It doesn't affect anything, just wondering.

File: src/plugin.py
Comment (FriendsOfGalaxy) [reply src/plugin.py:51]: I meant game ownership. Installation status is a different story. 

You probably meant "is owned", not "is installed"?
Because there is a lot of ways to find installed status like
- HKEY_CURRENT_USER_\SOFTWARE\Mojang\InstalledProducts\Minecraft Dungeons Launcher 
- HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall
- local files like  %appdata%/.minecraft_dungeons..

Comment (FriendsOfGalaxy) [reply src/plugin.py:51]: wow, this you just found it and it was created yesterday oO; nice lib.
If the web page knows if you have bought a game there is way to figure out ownership. I didn't bought it so can't help now.

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: why just not `return GameTime(game_id, time, lastPlayed)` without condition `time == 0 or lastPlayed is None:`?

is it somehow invalid scenario if only one of those values is None?

File: src/utils/time_tracker.py
Comment (FriendsOfGalaxy) [inline src/utils/time_tracker.py]: probably not used now?

**Review by FriendsOfGalaxy — APPROVED** [review]
Looks fine. @urwrstkn8mare I left some comments, but nothing important. I'll merge once get a chance to test it again.

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py:51]: > You probably meant "is owned", not "is installed"?

Yep. Sorry about that :) 

> wow, this you just found it and it was created yesterday

That's odd. It definitely wasn't created yesterday but for some reason, it says the repo was created yesterday with one commit? Very very weird. It still has the same full documentation and from a quick glance it looks like nothing has changed since I found it *more than a week ago* :). 

> If the web page knows if you have bought a game there is way to figure out ownership.

Yep, just gotta figure out how.

File: src/utils/time_tracker.py
Comment (urwrstkn8mare) [reply src/utils/time_tracker.py]: `GamesStillBeingTrackedException` is not needed anymore and can be removed. However, `GameNotTrackedException` is still used in `get_game_time()`.

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: Yeah, I think you're right I don't think it is necessary. I'll make this change and [this one](https://github.com/FriendsOfGalaxy/galaxy-integration-minecraft/pull/8#discussion_r490075587) quickly and add it to [the other pull request](https://github.com/TouwaStar/Galaxy_Plugin_Minecraft/pull/11). It won't be included in this release but it should be in the next release once more changes have been made as the current changes are minimal and don't really do anything.

File: src/plugin.py
Comment (FriendsOfGalaxy) [inline src/plugin.py]: One more thing I've found while testing old plugin - new version transition.

GOG user faces a unknown window (that has Galaxy-like CSS) on Galaxy start, that is unfair for Galaxy UX guys... Some users won't be interested in it at all and they will blame Galaxy for this "WTF" experience.

I have 2 solutions, you @urwrstkn8mare choose:
- (easy one) display ConnectionLost for all users. They need to click "Connect" again - now we have explicit. Technically: raise `InvalidCredentials` from `authenticate` if there is no `owned` in `stored_credentials`
- (nice one) `misc.get_next_step("Minecraft integration: new update", 720, 720, "migration_1.0")`. Page "zero" have only a few things:
    - Info that Minecraft integration has been updated and now need configuration.
    - (optionally) list of new features, but very short
    - buttons: Proceed (main button, will go to page1) | Logout (raise InvalidCredentials) | Skip (either to run the same page zero on next restart, or to function without new features (only Mincraft game owned??)

**Review by FriendsOfGalaxy — CHANGES_REQUESTED** [review]

File: src/plugin.py
Comment (urwrstkn8mare) [reply src/plugin.py]: > that is unfair for Galaxy UX guys

Sorry, I just used Galaxy-like CSS for consistency. I didn't mean to be unfair to anyone. 

> you @urwrstkn8mare choose

I think I would choose the nice one because first, it is nice :), and second it doesn't force users on to the new features/integration giving them a choice to stay with what they have. However, to implement it'll have to wait approximately a week or two. 

Sorry about that, I'll try to get it done as soon as possible.

Comment (FriendsOfGalaxy) [reply src/plugin.py]: @urwrstkn8mare OK good choice :) If the ability to stay in old flow is problematic for you in terms of time or whatever - feel free to go with the "easy one". Having those new features in general has more value than one-time inconvenience for users. If you have more questions for me - ask!

Comment (urwrstkn8mare) [reply src/plugin.py]: Thanks!

Comment (urwrstkn8mare) [reply src/plugin.py]: @FriendsOfGalaxy please see https://github.com/TouwaStar/Galaxy_Plugin_Minecraft/pull/13

**Review by FriendsOfGalaxy — APPROVED** [review]
🎉 
(sorry for late response)

**Pull request merged**
